vcd/resource_vcd_vapp_firewall_rules.go

```diff
@@ -0,0 +1,394 @@
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappId := d.Get("vapp_id").(string)
+	vapp, err := vdc.GetVAppById(vappId, false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %s", err)
+	}
+
+	vappNetwork, err := vapp.GetVappNetworkById(d.Get("network_id").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp network. %s", err)
+	}
+
+	var rules []map[string]interface{}
+	for _, rule := range vappNetwork.Configuration.Features.FirewallService.FirewallRule {
+		singleRule := make(map[string]interface{})
+		singleRule["description"] = rule.Description
+		singleRule["enabled"] = rule.IsEnabled
+		singleRule["policy"] = rule.Policy
+		singleRule["protocol"] = getProtocol(*rule.Protocols)
+		singleRule["destination_port"] = strings.ToLower(rule.DestinationPortRange)
+		singleRule["destination_ip"] = strings.ToLower(rule.DestinationIP)
+		if rule.DestinationVM != nil {
+			singleRule["destination_vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.DestinationVM.VAppScopedVMID)
+			singleRule["destination_vm_nic_id"] = rule.DestinationVM.VMNicID
+			singleRule["destination_vm_ip_type"] = rule.DestinationVM.IPType
+		}
+		singleRule["source_port"] = strings.ToLower(rule.SourcePortRange)
+		singleRule["source_ip"] = strings.ToLower(rule.SourceIP)
+		if rule.SourceVM != nil {
+			singleRule["source_vm_id"] = getVmIdFromVmVappLocalId(vapp, rule.SourceVM.VAppScopedVMID)
+			singleRule["source_vm_nic_id"] = rule.SourceVM.VMNicID
+			singleRule["source_vm_ip_type"] = rule.SourceVM.IPType
+		}
+		singleRule["enable_logging"] = rule.EnableLogging
+		rules = append(rules, singleRule)
+	}
+	_ = d.Set("rule", rules)
+	_ = d.Set("default_action", vappNetwork.Configuration.Features.FirewallService.DefaultAction)
+	_ = d.Set("log_default_action", vappNetwork.Configuration.Features.FirewallService.LogDefaultAction)
+
+	return nil
+}
+
+func getVmIdFromVmVappLocalId(vapp *govcd.VApp, vmVappLocalId string) string {
+	for _, vm := range vapp.VApp.Children.VM {
+		if vm.VAppScopedLocalID == vmVappLocalId {
+			return vm.ID
+		}
+	}
+	return ""
+}
+
+func expandVappFirewallRules(d *schema.ResourceData, vapp *govcd.VApp) ([]*types.FirewallRule, error) {
+	firewallRules := []*types.FirewallRule{}
+	for _, singleRule := range d.Get("rule").([]interface{}) {
+		configuredRule := singleRule.(map[string]interface{})
+
+		var protocol *types.FirewallRuleProtocols
+		// Allow upper and lower case protocol names
+		switch strings.ToLower(configuredRule["protocol"].(string)) {
+		case "tcp":
+			protocol = &types.FirewallRuleProtocols{
+				TCP: true,
+			}
+		case "udp":
+			protocol = &types.FirewallRuleProtocols{
+				UDP: true,
+			}
+		case "icmp":
+			protocol = &types.FirewallRuleProtocols{
+				ICMP: true,
+			}
+		case "tcp&udp":
+			protocol = &types.FirewallRuleProtocols{
+				TCP: true,
+				UDP: true,
+			}
+		default:
```

Comment:
From the schema description it is not clear what is the protocol default, and that an empty string results in "Any".
Should we make an explicit default in the schema?

Reply:
Added to schema default to "any"